packages/@ember/-internals/meta/lib/meta.ts

```diff
@@ -441,82 +478,136 @@ export class Meta {
+    if (i !== -1 && i < this._inheritedEnd) {
+      listeners.splice(i, 1);
+      this._inheritedEnd--;
+      i = -1;
     }
-    let pointer = this.parent;
-    while (pointer !== null) {
-      let listeners = pointer._listeners;
-      if (listeners !== undefined) {
-        this._listeners = this._listeners.concat(listeners);
-      }
-      if (pointer._listenersFinalized) {
-        break;
-      }
-      pointer = pointer.parent;
+
+    // if not found, push
+    if (i === -1) {
+      listeners.push({
+        event,
+        target,
+        method,
+        kind,
+      } as Listener);
+    } else {
+      // update own listener
+      listeners[i].kind = kind;
     }
-    this._listenersFinalized = true;
   }
 
-  removeFromListeners(eventName: string, target: any, method: Function | string): void {
-    let pointer: Meta | null = this;
-    while (pointer !== null) {
-      let listeners = pointer._listeners;
-      if (listeners !== undefined) {
-        for (let index = listeners.length - 4; index >= 0; index -= 4) {
-          if (
-            listeners[index] === eventName &&
-            (!method || (listeners[index + 1] === target && listeners[index + 2] === method))
-          ) {
-            if (pointer === this) {
-              listeners.splice(index, 4); // we are modifying our own list, so we edit directly
-            } else {
-              // we are trying to remove an inherited listener, so we do
-              // just-in-time copying to detach our own listeners from
-              // our inheritance chain.
-              this._finalizeListeners();
-              return this.removeFromListeners(eventName, target, method);
-            }
+  // check if we our meta is owned by a prototype
+  // if so, our listeners are inheritable so check if we have
+  // cached our flattened listeners, if so then clear the inherited listeners
+  // and bump the global version count
+  private writableListeners(): Listener[] {
+    let listeners = this._listeners;
+
+    if (listeners === undefined) {
+      listeners = this._listeners = [] as Listener[];
+    }
+
+    if (this.source === this.proto && this._inheritedVersion > 0) {
+      currentListenerVersion++;
+    }
+
+    return listeners;
+  }
+
+  protected flattenedListeners(): Listener[] | undefined {
+    let parent = this.parent;
+
+    if (parent !== null && this._inheritedVersion < currentListenerVersion) {
+      // compute
+      let parentListeners = parent.flattenedListeners();
+
+      if (parentListeners !== undefined) {
+        let listeners = this._listeners;
+
+        if (listeners === undefined) {
+          listeners = this._listeners = [] as Listener[];
+        }
+
+        for (let i = 0; i < parentListeners.length; i++) {
+          let listener = parentListeners[i];
+          let index = indexOfListener(listeners, listener.event, listener.target, listener.method);
+
+          if (index === -1) {
+            listeners.unshift(listener);
```

Comment:
The current logic relies on the fact that we are updating objects that are shared, but this doesn't work in the `REMOVE_ALL` case. We'll need to splice out any inherited listeners in the event of an update to make sure we're updating everything correctly